Fix master-next build (Part 2)

MForster opened the pull request:
These changes fix the `master-next` build together with some other changes that I'm submitting as pull request https://github.com/apple/swift/pull/31089 against `master`.

MForster commented:
Note that I'm happy to reformat the `llvm::interleave` changes. But I think it's better to do that when this gets applied to `master` to avoid merge conflicts.

gribozavr commented on include/swift/AST/SourceFile.h:
```diff
@@ -734,9 +734,11 @@ struct DenseMapInfo<swift::SourceFile::ImportedModuleDesc> {
-    return combineHashValue(ImportedModuleDMI::getHashValue(import.module),
-           combineHashValue(ImportOptionsDMI::getHashValue(import.importOptions),
-                            StringRefDMI::getHashValue(import.filename)));
+    return detail::combineHashValue(
```
I'm not sure if functions from the `detail` namespace are supposed to be called here.

gribozavr replied:
Scratch that, I don't think there is a better way. LLVM is indeed only using it as an implementation detail, but Swift had its own copy which is now deleted.

gribozavr commented on include/swift/Basic/STLExtras.h:
```diff
@@ -71,40 +71,6 @@ struct function_traits<R (T::*)(Args...) const> {
-inline void interleave(const Container &c, UnaryFunctor each_fn,
-                       NullaryFunctor between_fn) {
-  interleave(c.begin(), c.end(), each_fn, between_fn);
-}
```
Could we merge this change into `master` instead? Move this function into the LLVM namespace in this file, and then update all callers to use `llvm::` qualification. Then on the `master-next` branch, only delete the redundant implementation.

MForster replied:
See pull request https://github.com/apple/swift/pull/31091.

MForster replied:
I've removed the `llvm::interleave` changes from this pull request. Let's continue any discussion about this on https://github.com/apple/swift/pull/31091.

gribozavr commented:
@swift-ci Please test

MForster committed:
Remove Swift's copy of llvm::interleave

This is only needed on the master branch, which doesn't have LLVM's
copy, yet.

MForster committed:
Use LLVM version of combineHashValue

This fixes the master-next build.

MForster committed:
Call the new BinaryOperator factory function

The constructor is not public any more. This fixes the master-next build.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

MForster committed:
Ignore ExtInt for C interop and IRGen

This feature was introduced here:
https://github.com/apple/llvm-project/commit/61ba1481e200b5b35baa81ffcff81acb678e8508

MForster commented:
A few more breakages had arrived over the weekend. I've added another string fix to `master` (https://github.com/apple/swift/pull/31148) and another commit (https://github.com/apple/swift/pull/31090/commits/d416258aabf4844f87df907e74736a7486455e3e) to this pull request.

gribozavr commented:
@swift-ci Please test

gribozavr commented on lib/IRGen/GenCall.cpp:
```diff
@@ -700,7 +700,7 @@ namespace {
       case clang::Type::IncompleteArray:
       case clang::Type::VariableArray:
-        llvm_unreachable("variable-sized or incomplete array in ABI lowering");
+        llvm_unreachable("variable-sized or incompleteincomplete array in ABI lowering");
```
Unintended edit?

MForster replied:
Oops. Fixed.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

xedin commented:
Going to force merge since it fixes some of the problems currently present in master-next.
